hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java

```diff
@@ -337,10 +505,19 @@ private void uploadBlockAsync(DataBlocks.DataBlock blockToUpload,
              * leaseId - The AbfsLeaseId for this request.
              */
             AppendRequestParameters reqParams = new AppendRequestParameters(
-                offset, 0, bytesLength, mode, false, leaseId, isExpectHeaderEnabled);
-            AbfsRestOperation op = getClient().append(path,
-                blockUploadData.toByteArray(), reqParams, cachedSasToken.get(),
-                contextEncryptionAdapter, new TracingContext(tracingContext));
+                offset, 0, bytesLength, mode, false, leaseId,
+                isExpectHeaderEnabled);
+            AbfsRestOperation op;
+            try {
+              op = remoteWrite(blockToUpload, blockUploadData, reqParams,
+                  tracingContext);
+            } catch (InvalidIngressServiceException ex) {
+              switchHandler();
+              // retry the operation with switched handler.
+              op = remoteWrite(blockToUpload, blockUploadData,
```

Comment:
At this moment, the handler has been switched, but the blockToUpload can be of non-compatible class. We have a hard_casting in handler-> remoteWrite(), the blobIngressHandler would cast to AbfsBlobBlock. We might need to check if the block has to be switched as well.

Reply:
No tested this with the test testCreateOverDfsAppendOverBlob, the blovkToUpload is of kind AbfsBlock which is a parent to AbfsBlobBlock and hence no switching needed

Reply:
okay, looks like conversion from dfsIngressHandler to blobIngressHandler cannot happen in a valid config environment. But, would like to discuss one thing here at https://github.com/ABFSDriver/AbfsHadoop/blob/f8483e801a1cbd5d9fa02ac109550a31d0137d3d/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AzureBlobIngressHandler.java#L118, we have a hard-casting to AbfsBlobBlock.   Due to the config constraint, it is not a problem which is good, but can we have a way to remove hard-casting.

Reply:
if the hard casting is not done, we wont be able to access functions like getBlockId. I am not able to get the concern fully.

Reply:
I understand by the means of config we say that DFSIngress can not convert to BlobIngress, and if it happens, we will cite that its config issue (fs.azure.enable.dfstoblob.fallback is not properly configured).

There is a code block somewhere outside which ensures that casting problem does not happen, but if we see at the class level and for code-readablity it becomes confusing. Also, it gets flagged by spotbugs (so for OSS, we would have to remove the hard-casting):
```
<file classname="org.apache.hadoop.fs.azurebfs.services.AzureBlobIngressFallbackHandler">
<BugInstance type="BC_UNCONFIRMED_CAST" priority="Normal" category="STYLE" message="Unchecked/unconfirmed cast from org.apache.hadoop.fs.azurebfs.services.AbfsBlock to org.apache.hadoop.fs.azurebfs.services.AbfsBlobBlock in org.apache.hadoop.fs.azurebfs.services.AzureBlobIngressFallbackHandler.bufferData(AbfsBlock, byte[], int, int)" lineNumber="86"/>
</file>
```

I have a suggestion, can we blockId an abstract method in AbfsBlock, for the super class, we thow IlleagalException, and for the child class AbfsBlobBlock, we return the required blockId.

We can move entirely away from casting.

Reply:
taken